### ISSUE TITLE
Avoid unaligned access when enumerating addresses in the Rsdt or Xsdt ACPI table

### ISSUE THREAD

**Issue by kevinaboos** [theseus-os/Theseus#355]
### Problem ###

When building with debug assertions enabled, the strict alignment checks in `slice::from_raw_parts()` trigger with the panic message: 
> attempt to create unaligned or null slice

Some log statements I added to demonstrate the issue:
```
[D] kernel/acpi_table/src/lib.rs:225: signature [82, 83, 68, 84], loc TableLocation { offset: 1042, slice_offset_and_length: Some((1078, 3)) }, offset 1078, len 3
[D] kernel/memory/src/paging/mapper.rs:683: MappedPages::as_slice(): requested slice of type u32 (align 4) with length 3 (total size 12) at byte_offset 1078, MappedPages start 0x6A36000, size 4096!
[T] kernel/panic_wrapper/src/lib.rs:32: at top of panic_wrapper: PanicInfo { payload: Any, message: Some(attempt to create unaligned or null slice), location: Location { file: "/home/kevin/.rustup/toolchains/nightly-2020-11-10-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/slice/raw.rs", line: 90, col: 5 } }
```
The above offset of `1078` is unaligned with respect to the `u32` type, which is what causes the error when trying to parse the list of physical addresses at the end of the RSDT table.

Backtrace for posterity:
```
0xFFFFFFFF80AC06DB in core::panicking::panic::hb3564ca2785ca732 + 0x6B
0xFFFFFFFF80554E1B in core::slice::raw::from_raw_parts::hce58f8396369234e + 0x6B
0xFFFFFFFF805573C1 in memory::paging::mapper::MappedPages::as_slice::h097c539eaea2434f + 0xAE1
0xFFFFFFFF805515EB in acpi_table::AcpiTables::table_slice::h05304a51a0be57c7 + 0x55B
0xFFFFFFFF805587F4 in rsdt::RsdtXsdt::get::hfb908e227e9af5bb + 0x64
0xFFFFFFFF80535B7C in acpi::init::he6ecc1f71a14d6e8 + 0x75C
0xFFFFFFFF8035D119 in device_manager::early_init::h37837ff5564b989b + 0x79
0xFFFFFFFF8013851D in captain::init::h35dd2c8c63034ba7 + 0x1ED
0xFFFFFFFF8011DF33 in nano_core_start + 0x18F3
```

### Solution ###
Change the type definitions for `Rsdt` and `Xsdt` to store their dynamic slice of addresses as `[u8]` instead of `[u32]` and `[u64]` respectively, because the `u8` type has an alignment requirement of 1 byte (obviously).  Currently they are defined as such:
```rust
type Rsdt<'t> = (&'t Sdt, &'t [u32]);
type Xsdt<'t> = (&'t Sdt, &'t [u64]);
```

Then, in the `RsdtXsdt::addresses()` function, return a new iterator-like struct, in which that struct's `Iterator::next()` function creates each physical addresses from the current index into the `[u8]` on demand.

**Comment by kevinaboos** [comment]
TODO: after this is done, explore whether we would like to enforce alignment requirements for the type parameter `T` in `MappedPages::as_slice()` and `MappedPages::as_slice_mut()`

**Comment by kevinaboos** [comment]
Closed by 25fa14d

**Comment by kevinaboos** [comment]
> TODO: after this is done, explore whether we would like to enforce alignment requirements for the type parameter `T` in `MappedPages::as_slice()` and `MappedPages::as_slice_mut()`

In the current version of Theseus, there are no other unaligned memory accesses for slices, so we _could_ enforce this if we wanted to. But it's not necessary at the moment.
